Mathlib/CategoryTheory/Limits/FormalCoproducts.lean

```diff
@@ -0,0 +1,428 @@
+construct the pullback in `FormalCategory.{w} C`. -/
+def pullbackCone : PullbackCone f g :=
+  .mk (W := ⟨Function.Pullback f.f g.f, fun i ↦ (pb i).pt⟩)
+    ⟨fun i ↦ i.1.fst, fun i ↦ (pb i).fst⟩
+    ⟨fun i ↦ i.1.snd, fun i ↦ (pb i).snd⟩
+    (hom_ext (funext fun i ↦ i.2) (fun i ↦ by simp [(pb i).condition]))
+
+section simp_lemmas
+
+@[simp] lemma pullbackCone_fst_f (i) : (pullbackCone f g pb).fst.f i = i.1.1 := rfl
+
+@[simp] lemma pullbackCone_fst_φ (i) : (pullbackCone f g pb).fst.φ i = (pb i).fst := rfl
+
+@[simp] lemma pullbackCone_snd_f (i) : (pullbackCone f g pb).snd.f i = i.1.2 := rfl
+
+@[simp] lemma pullbackCone_snd_φ (i) : (pullbackCone f g pb).snd.φ i = (pb i).snd := rfl
+
+@[simp] lemma pullbackCone_condition :
+    (pullbackCone f g pb).fst ≫ f = (pullbackCone f g pb).snd ≫ g :=
+  PullbackCone.condition _
+
+end simp_lemmas
+
+/-- The `Equiv` that witnesses that `pullbackCone f g pb` is actually a pullback. This is the
+universal property of pullbacks. -/
+@[simps!] def homPullbackEquiv : (T ⟶ (pullbackCone f g pb).pt) ≃
+    { p : (T ⟶ X) × (T ⟶ Y) // p.1 ≫ f = p.2 ≫ g } where
+  toFun m := ⟨⟨m ≫ (pullbackCone f g pb).fst, m ≫ (pullbackCone f g pb).snd⟩, by simp⟩
+  invFun s := ⟨fun i ↦ ⟨(s.1.1.f i, s.1.2.f i), congrFun (congrArg Hom.f s.2) i⟩,
+    fun i ↦ (hpb _).lift (PullbackCone.mk (s.1.1.φ i) (s.1.2.φ i)
+      (by simpa using ((hom_ext_iff _ _).1 s.2).2 i))⟩
+  left_inv m := hom_ext rfl (fun i ↦ by
+    simp only [category_comp_f, category_comp_φ, eqToHom_refl, Category.comp_id]
+    exact (hpb _).hom_ext ((pb _).equalizer_ext (by aesop) (by aesop)))
+  right_inv s := by ext <;> simp
+
+/-- `pullbackCone f g pb` is a pullback. -/
+def isLimitPullbackCone : IsLimit (pullbackCone f g pb) := by
+  refine PullbackCone.IsLimit.mk
+    (fst := (pullbackCone f g pb).fst) (snd := (pullbackCone f g pb).snd) _
+    (fun s ↦ (homPullbackEquiv f g pb hpb s.pt).2 ⟨(s.fst, s.snd), s.condition⟩)
+    (fun s ↦ congrArg (·.1.fst)
+      ((homPullbackEquiv f g pb hpb s.pt).right_inv ⟨(s.fst, s.snd), s.condition⟩))
+    (fun s ↦ congrArg (·.1.snd)
+      ((homPullbackEquiv f g pb hpb s.pt).right_inv ⟨(s.fst, s.snd), s.condition⟩))
+    (fun s m h₁ h₂ ↦ ?_)
+  convert ((homPullbackEquiv f g pb hpb s.pt).left_inv m).symm using 3
+  rw [← h₁, ← h₂]; rfl
+
+-- Arguments cannot be inferred.
+include pb hpb in
+theorem hasPullback_of_pullbackCone : HasPullback f g :=
+  ⟨⟨⟨_, isLimitPullbackCone f g pb hpb⟩⟩⟩
+
+omit pb
+variable [HasPullbacks C]
+
+instance : HasPullback f g :=
+  hasPullback_of_pullbackCone f g (fun _ ↦ pullback.cone _ _) (fun _ ↦ pullback.isLimit _ _)
+
+instance : HasPullbacks (FormalCoproduct.{w} C) :=
+  hasPullbacks_of_hasLimit_cospan _
+
+include pb
+
+/-- The arbitrary choice of pullback is isomorphic to the explicitly constructed pullback
+`pullbackCone f g pb`. -/
+noncomputable def pullbackIsoPullbackConePt : pullback f g ≅ (pullbackCone f g pb).pt :=
+  limit.isoLimitCone ⟨_, isLimitPullbackCone f g pb hpb⟩
```

Comment:
I am not completely convinced that this API is really necessary.

Reply:
What should I replace it with? How would I prove properties about an arbitrary `pullback`? Or are you saying we shouldn't use it to begin with? But surely there are theorems that are explicitly stated in terms of this arbitrary `pullback`...

Reply:
The important definition is `isLimitPullbackCone`. (It would be nice also to add a `IsPullback` lemma.)
Otherwise, I do not think there is a need to define an isomorphism between a certain (co)cone point and the (co)limit everytime we show that a certain (co)cone is (co)limit. I believe these definitions should be removed.

Reply:
done